docs: first pass at documenting how to add extensions (#11540)
envoyproxy/envoy: CONTRIBUTING.md

* If a PR includes a deprecation/breaking change, notification should be sent to the
  [envoy-announce](https://groups.google.com/forum/#!forum/envoy-announce) email list.

# Adding new extensions

> For developers adding a new extension, one can largely model off of existing extensions.

Reviewer: ```suggestion
For developers adding a new extension, one can take an existing extension as the starting point.
```

Extension configuration should be located in a directory structure like
`api/envoy/extensions/area/plugin/`, for example `api/envoy/extensions/tracers/zipkin/`

The code for the extension should be located under the equivalent
`source/extensions/area/plugin`, and include an envoy_cc_extension with the
configuration and tagged with the appropriate security posture, and an

> envoy_cc_library with the code.

Reviewer: Suggestion: use markup to make Bazel macros distinct, e.g. `*envoy_cc_library*`.

> 

Reviewer: Link to https://github.com/envoyproxy/envoy/blob/master/api/STYLE.md#adding-an-extension-configuration-to-the-api

Other changes will likely include

>   * Editing source/extensions/extensions_build_config.bzl to include the new extensions

Reviewer: Suggestion: use `:repo:` to turn these into real links.

* If a PR includes a deprecation/breaking change, notification should be sent to the
  [envoy-announce](https://groups.google.com/forum/#!forum/envoy-announce) email list.

# Adding new extensions

For developers adding a new extension, one can take an existing extension as the starting point.

Extension configuration should be located in a directory structure like

> `api/envoy/extensions/area/plugin/`, for example `api/envoy/extensions/tracers/zipkin/`

Reviewer 2: Let's use other extension as examples other than tracers, they only have `v4alpha` which might be confusing.

The code for the extension should be located under the equivalent
`source/extensions/area/plugin`, and include an *envoy_cc_extension* with the
configuration and tagged with the appropriate security posture, and an
*envoy_cc_library* with the code. More details on how to add a new extension

> API can be found [here](api/STYLE.md):

Reviewer: ```suggestion
API can be found [here](api/STYLE.md#adding-an-extension-configuration-to-the-api):
```
